### ISSUE TITLE
Ellipsoid Factor is very slow

### ISSUE THREAD

**Issue by mdoube** [bonej-org/BoneJ2#188]
**Describe the bug**
Ellipsoid Factor in BoneJ2 is about 1-2 orders of magnitude slower than in BoneJ1. It also seems to be a memory hog, using 10s of GB of RAM given the chance.

Found 6219 ellipsoids in 542822 ms (BoneJ2)
Found 6218 ellipsoids in 11904 ms (BoneJ1)

Found maximal ellipsoids in 53371 ms (BoneJ2)
Found maximal ellipsoids in 634 ms (BoneJ1)

This may reflect poor big-O scaling - the test image above used only every 25th skeleton point in a 392×392×393 pixel volume (so, not very big).

Importantly, this is the only moderately refactored legacy EF1 code.

**To Reproduce**
1. Run Ellipsoid Factor in BoneJ2 and BoneJ1 (legacy ImageJ) on the same image with the same settings

**Expected behavior**
Timings should be approximately the same

**Additional context**
Use of Streams and Lambda expressions seems the likely culprit.
This won't need any work if EF2 is released soon.

**Comment by alessandrofelder** [comment]
What are the parameters? Which version are you running?
emu.tif is pretty comparable in timings on my machine between BoneJ1 and BoneJ2.
[this is the branch I am working on currently](https://github.com/bonej-org/BoneJ2/tree/refactor-EF2)
My experience is that emu.tif takes 180 seconds in both cases,
Running with points/ellipsoid = 5 and seeds only on distance ridge (not on surface), 100 directions and other standard parameters for EF1

**Comment by mdoube** [comment]
I'm running the last commit that built successfully, which is #185 bonej-org/master 
https://github.com/bonej-org/BoneJ2/commit/9aa1b9335065544ef453c736ce3d389943df76d0

Using emu.tif I get for BoneJ2 (in Fiji)
Found 7477 ellipsoids in 29154 ms
Found maximal ellipsoids in 6663 ms

for BoneJ1 (in IJ1)
Found 7490 ellipsoids in 14505 ms
Found maximal ellipsoids in 496 ms

So that's a 2× speed difference for ellipsoid fitting and a 12× speed difference for filtering. There is probably a scaling problem as ellipsoids get bigger. This is on a 40-core dual CPU workstation. It also seems like the BoneJ2 version doesn't max out the CPU cores in the same way that the BoneJ1 version does. There were some changes to the multithreading implementation in the [file history](https://github.com/bonej-org/BoneJ2/commits/master/Legacy/bonej/src/main/java/org/bonej/plugins/EllipsoidFactor.java).

**Comment by alessandrofelder** [comment]
Ah. I see. Early in the morning here, misunderstood the first post :) Thanks for clarifying. Yes, let's release EF2 soon!

**Comment by mdoube** [comment]
On `umzc_378p_Apteryx_hasstii_head.tif`:

BoneJ1 / EF1 / Vanilla ImageJ:
```
Found 1564 ellipsoids in 5223 ms
Found maximal ellipsoids in 879 ms
```
BoneJ2 / EF2 #209 / Fiji is Just ImageJ
```
[INFO] Found 1571 ellipsoids in 133636 ms
[INFO] Found maximal ellipsoids in 12753 ms
```

**Comment by mdoube** [comment]
On `shrew.tif` and a 4-thread laptop, all skeleton points used:

EF2
```
[INFO] Found 2663 skeleton-seeded ellipsoids.
[INFO] Found 2663 ellipsoids in 123987 ms
[INFO] Found maximal ellipsoids in 2892 ms
```
EF1
```
Found 2712 skeleton points
Found 2360 ellipsoids in 21906 ms
Found maximal ellipsoids in 322 ms
```

**Comment by mdoube** [comment]
`Apteryx_haastii` on laptop, every 100th point:
EF2
```
[INFO] Found 1571 skeleton-seeded ellipsoids.
[INFO] Found 1571 ellipsoids in 206571 ms
[INFO] Found maximal ellipsoids in 88367 ms
```
EF1
```
Found 1557 ellipsoids in 45241 ms
Found maximal ellipsoids in 2013 ms
```
Also looks like the zipped image is corrupted, will rezip when next at workstation. Comparing EF1 between workstation (40 threads, 45 s) and laptop (4 threads, 5 s) suggests that there is ~linear scaling with n  threads. A less than double speedup exists for EF2, despite 10× more threads available. Likely culprit a poor multithreading implementation. Would be worth checking CPU utilisation on multi-thread machine.

**Comment by mdoube** [comment]
Oh dearie me... jvisualvm sampler attached to ellipsoid fitting:
![image](https://user-images.githubusercontent.com/87103/76317609-6d74f300-6317-11ea-83ba-c93701e2f8d9.png)

**Comment by alessandrofelder** [comment]
Hmmm interesting :( on RVC, 24-core machine, I get
EF2
```
Found 42429 ellipsoids in 148523 ms
```
EF1
```
44334 pixels unfilled with ellipsoids out of 1878652 total foreground pixels
```

**Comment by alessandrofelder** [comment]
inlining unit sphere random vector generator instead of using apache math3 should be straightforward to do, of course...

**Comment by alessandrofelder** [comment]
also, thanks for doing all this validation work, @mdoube

**Comment by rimadoma** [comment]
@alessandrofelder Could try another way generating a vector with a random orientation, as in `Vectors.randomVector()` in `Legacy`

**Comment by mdoube** [comment]
emu.tif, EF2, 40-core workstation, today's ellipsoid-factor-master:
```
[INFO] Found 42043 skeleton-seeded ellipsoids.
[INFO] Found 42043 ellipsoids in 140179 ms
[INFO] Found maximal ellipsoids in 8076 ms
```

**Comment by mdoube** [comment]
EF1 profiling:
![image](https://user-images.githubusercontent.com/87103/76402030-ca2ae900-63bd-11ea-9fd5-e2ff7f773a79.png)

**Comment by alessandrofelder** [comment]
On 4 core laptop, shrew.tif:
EF2:
```
[INFO] Found 2663 skeleton-seeded ellipsoids.
[INFO] Found 2663 ellipsoids in 21528 ms
[INFO] Found maximal ellipsoids in 874 ms
```
EF1
```
Found 2407 ellipsoids in 4858 ms
Found maximal ellipsoids in 321 ms
```

**Comment by alessandrofelder** [comment]
Observation: if we want our search vectors to be distributed uniformly on the ellipsoid surface, we need a variable number of random vectors distributed on the sphere to achieve this.

**Comment by alessandrofelder** [comment]
An attempt at speeding up the surface point finding is here: [https://github.com/bonej-org/BoneJ2/tree/improve-speed-surface-point-sampling](https://github.com/bonej-org/BoneJ2/tree/improve-speed-surface-point-sampling)

It reduces my shrew.tif execution time significantly, but the loss of variation in the random sampling leads to some edge artefacts :(

**Comment by mdoube** [comment]
Just realised the fundamental difference in approaches is that EF2 achieves random sampling over the ellipsoid, which is different for every iteration, whereas EF1 only achieves random sampling over the unit sphere, which can be reused for each iteration but which introduces variable sampling density over the ellipsoid.

For most iterations the ellipsoid doesn't change very much, so perhaps the random vectors could be recalculated only every 10th step or when something important happens (e.g. Background contact)

**Comment by mdoube** [comment]
> random sampling over the ellipsoid, which is different for every iteration

The net effect of which is a sort of 'twinkling' of sampling spots as the ellipsoid goes through its small (sort-of-Brownian) bump-turn-wiggle-dilate-contract movements. Blurred in time it's a good way to approximate the true ellipsoid surface and avoid sampling artefacts.

Given that it is expensive to calculate new random points for every iteration, it would be worth putting some thought into how to do that more efficiently, like, only for the first dilate, or only for moves that change the aspect ratio of the ellipsoid, or only every 10th move, or... then reusing the unit vectors for the next 9 moves.

**Comment by alessandrofelder** [comment]
Yes, my thoughts were going in a similar direction. Will have a go at implementing now :)

**Comment by mdoube** [comment]
emu.tif 4-core laptop, EF2, settings as above. Will try soon on office workstation.
```
[INFO] Found 8405 skeleton-seeded ellipsoids.
[INFO] Found 8405 ellipsoids in 108703 ms
[INFO] Found maximal ellipsoids in 21551 ms
```
EF1:
```
Found 8251 ellipsoids in 74263 ms
Found maximal ellipsoids in 2010 ms
` ``

**Comment by mdoube** [comment]
Latest commit helps but random point calculation is still problematic:
![Screenshot from 2020-03-13 17-16-19](https://user-images.githubusercontent.com/87103/76607198-84e9f100-654e-11ea-9970-535b197643dc.png)

Expensive methods / hot spots:
`java.util.Random.nextGaussian()` in `wiggle()`
Does it really have to be Gaussian, and could it be precomputed if so, e.g. a list of 1000 values from a Gaussian distribution from which you pick at random?

`getAxisAlignRandomlyDistributedSurfacePoints()` is still expensive: is it the calculation of the points or waiting to read them from the image? (i.e. memory access latency).

**Comment by mdoube** [comment]
> emu.tif 4-core laptop, EF2, settings as above

Today:
```
[INFO] Found 8421 skeleton-seeded ellipsoids.
[INFO] Found 8421 ellipsoids in 209053 ms
[INFO] Found maximal ellipsoids in 43149 ms
```

**Comment by mdoube** [comment]
> emu.tif, EF2, 40-core workstation, today's ellipsoid-factor-master on Fiji
> 
> ```
> [INFO] Found 42043 skeleton-seeded ellipsoids.
> [INFO] Found 42043 ellipsoids in 140179 ms
> [INFO] Found maximal ellipsoids in 8076 ms
> ```
> 
> EF1, 40-core workstation, on vanilla ImageJ:
> 
> ```
> Found 41278 ellipsoids in 41242 ms
> Found maximal ellipsoids in 4482 ms
> ```

Today, EF2, 40-thread workstation, [emu.tif](http://bonej.org/images/emu.tif)
```
[INFO] Found 42045 skeleton-seeded ellipsoids.
[INFO] Found 42045 ellipsoids in 102854 ms
[INFO] Found maximal ellipsoids in 11891 ms
```

**Comment by mdoube** [comment]
Let's close this issue, as it's largely addressed by changes in #209 

There is definitely still room to take lessons from experience so far, especially with regard to detecting hotspots and reducing them. Efficient generation of sampling points on the ellipsoid surface seems to be the main bottleneck.
